### DISCUSSION TITLE
SignaturePad 签名组件在开发环境可以正常工作，在生产服务器就显示哭脸和大片空白

### DISCUSSION THREAD

**Discussion by WeiJunFenYou** [dotnetcore/BootstrapBlazor#4374] [Q&A]
`    <Modal @ref="PadModal">
        <ModalDialog IsCentered="true" Title="请签名" ShowSaveButton="false" SaveButtonText="确认" Size="Size.Medium" >
            <BodyTemplate>
                <SignaturePad OnResult="OnResult" />
            </BodyTemplate>
        </ModalDialog>
    </Modal>`

代码如上所示，在本地调试的时候，正常工作：
![image](https://github.com/user-attachments/assets/6d9e2c64-ff3f-4bde-a7c2-1aa2c953e032)

当发布到生产服务器时，就变成了这样：
![image](https://github.com/user-attachments/assets/efbda1ee-a860-42a2-ab41-637f336bbe1a)

**Comment by WeiJunFenYou** [comment]
这可能是资源文件没有加载成功。生产环境的网络和开发环境有异，才导致的。不知这种情况要如何解决？

**Comment by WeiJunFenYou** [comment]
查了一下html，开发环境中，页面有一个css文件
![image](https://github.com/user-attachments/assets/477d9c90-543b-4836-a1bf-0b6618ac6b63)

但在生产环境没有

<img width="306" alt="image" src="https://github.com/user-attachments/assets/2a36b6ef-b2a0-4223-be1b-dec86fd7ae80">

不知为何会缺失这个css文件

**Comment by WeiJunFenYou** [comment]
@ArgoZhang 麻烦大神看下

**Comment by WeiJunFenYou** [comment]
刚刚试了下，把BootstrapBlazor.SignaturePad却换到7.x的低版本后，在生产服务器也可以正常显示了。但7.x自带的无法手写的bug也就出现了，仍然无法使用这个组件。

初步确认css资源在生产服务器无法加载，是组件的问题。

**Reply by ArgoZhang** [reply]
@WeiJunFenYou 您的描述我无法理解出更多的信息，官网正常吗？我是不是可以理解官网也是发布的？能否比对一下差异在哪里？那个样式是自动生成的。

**Reply by WeiJunFenYou** [reply]
官网是正常的。官网使用的是最新版8.0.2的吗？官网和我的本地环境运行的写字板看起来并没有什么差异。

**Reply by ArgoZhang** [reply]
对的。官网就是 `8.0.2`

**Reply by WeiJunFenYou** [reply]
请问官网部署使用的是什么服务器呢？我使用的是iis，怀疑可能是iis配置的问题。

**Reply by ArgoZhang** [reply]
debian 11 + nginx 不是 windows 系统

**Reply by WeiJunFenYou** [reply]
这可能是组件的BUG，只在IIS服务器上部署时才会出现。

原因是：
1、使用8.0.2版本（最新版），签字板在IIS上可以正常显示（但无法使用，老版本的BUG），使用8.0.1以以下，就不能正常显示。
2、使用VS自带本地服务器IISExpress，以及nginx 服务器，任何版本都可以正常显示。

可能是组件在更新到8.0.2的时候，因为某些未知的改动，导致在IIS上CSS加载不正确。

**Reply by ArgoZhang** [reply]
请帮忙做个组件的示例，能够复现你说的条件，我还没有能力做出这种组件，需要如何编写代码才能做到部署到 IIS 中才出你所谓的 bug

**Reply by j4587698** [reply]
组件不可能出现什么在IIS上部署才能出现的BUG，这是个UI库，所有的内容呈现都是浏览器做的。如果你认为缺了css，检查一下你的发布文件里有没有这个css，检查一下你的F12里有没有浏览器的错误，然后访问以下那个url看下是否能找到这个css。
一步一步来，别上来就是组件BUG

**Reply by ArgoZhang** [reply]
@WeiJunFenYou 可以联系我们，有偿技术支持服务

**Reply by WeiJunFenYou** [reply]
> 组件不可能出现什么在IIS上部署才能出现的BUG，这是个UI库，所有的内容呈现都是浏览器做的。如果你认为缺了css，检查一下你的发布文件里有没有这个css，检查一下你的F12里有没有浏览器的错误，然后访问以下那个url看下是否能找到这个css。 一步一步来，别上来就是组件BUG

1、发布文件有这个css，但使用谷歌浏览器打开后，Sources里没有组件的css。
2、浏览器按下F12后，网络和控制台没有任何错误。
3、浏览器直接访问那个css，可以访问到。

**Reply by WeiJunFenYou** [reply]
有偿技术支持服务收费如何？

**Reply by j4587698** [reply]
首先，F12没有任何错误的话证明所有的加载应该都是成功的，失败了一般会有404或者500之类的错误提示。
css js这些加载都是由浏览器执行的，不管是IIS还是Nginx还是你直接启动，对浏览器都没有影响。
这种情况下跟组件库有啥关系啊？
你自己换个浏览器试试，换个电脑试试。为啥上来就怀疑组件库？

**Reply by WeiJunFenYou** [reply]
用排除法的确是跟组件有关，是不是BUG两说，至少组件是导致此现象的一个因素。因为低版本的组件没有这个问题。具体我再看以下组件的代码，在升级的过程中都发生了什么。后面如果找到原因我会在下面更新。

**Reply by WeiJunFenYou** [reply]
> 首先，F12没有任何错误的话证明所有的加载应该都是成功的，失败了一般会有404或者500之类的错误提示。 css js这些加载都是由浏览器执行的，不管是IIS还是Nginx还是你直接启动，对浏览器都没有影响。 这种情况下跟组件库有啥关系啊？ 你自己换个浏览器试试，换个电脑试试。为啥上来就怀疑组件库？

因为更换了旧版组件就没有这个问题了。

**Reply by ArgoZhang** [reply]
@WeiJunFenYou 300 起步，超过 1 小时另外算，建议挂上源码先确保本地无问题

**Reply by j4587698** [reply]
低版本没问题你考虑过没有是浏览器有缓存呢？你换过电脑吗？浏览器吗？清理过缓存吗？你这排除法什么都没换过怎么就把自己排除了？

**Reply by WeiJunFenYou** [reply]
> 低版本没问题你考虑过没有是浏览器有缓存呢？你换过电脑吗？浏览器吗？清理过缓存吗？你这排除法什么都没换过怎么就把自己排除了？

1、清理过缓存了。清理之后仍然如此。
2、换浏览器了。Chrome和Edge，更换浏览器仍然如此。
3、换电脑了。另一台电脑打开仍然如此。

**Reply by ArgoZhang** [reply]
我想不通为啥就非要一直说问题的结果，为啥就不上传一个复现工程？涉密项目吗？可以用模板做一个复现工程，方便我们排查问题。不要一直做无效的沟通，提高沟通效率最简单的办法就是提交复现工程

**Comment by WeiJunFenYou** [comment]
这个问题刚刚解决了。

我拉取了SignaturePad的源代码，经过修改后，自己打包编译成nuget包。具体过程是：

在SignaturePad组件中，删除了SignaturePad.razor.css，并将其替换为外联样式表。打包后项目就正常了。

SignaturePad.razor.css这种类型的样式表，是Blazor的实验性功能：CSS隔离。这大概是微软的锅。

感谢作者大大的耐心回复。给您填了堵。这个问题后续我会继续跟进，如果这种CSS隔离样式表真的存在一些问题，以后使用是要注意的。

**Reply by WeiJunFenYou** [reply]
@ArgoZhang

**Reply by ArgoZhang** [reply]
@WeiJunFenYou 大概率是你工程的问题。这么说吧，如果像你这样解决问题，那官网怎么就好使呢？

但是问题肯定是样式隔离导致的

**Reply by WeiJunFenYou** [reply]
是的。现在的解决方案治标不治本，只是临时凑合一下。所以我才说后面要继续跟进，找到根本原因。否则以后作者大大再发布类似的组件，我还要自己拉取源代码打包，那就不好了。😏

**Reply by ArgoZhang** [reply]
@WeiJunFenYou 我刚才想到一个问题点，应该是这个问题导致的

我猜你引用 SignaturePad 的工程未开启样式隔离，导致最终的样式隔离文件缺少 `import` 语句导致的

我们项目的样式文件如下：
```
@import '_content/BootstrapBlazor.AntDesignIcon/BootstrapBlazor.AntDesignIcon.bundle.scp.css';
@import '_content/BootstrapBlazor.Bluetooth/BootstrapBlazor.Bluetooth.bundle.scp.css';
@import '_content/BootstrapBlazor.CodeEditor/BootstrapBlazor.CodeEditor.bundle.scp.css';
@import '_content/BootstrapBlazor.ElementIcon/BootstrapBlazor.ElementIcon.bundle.scp.css';
@import '_content/BootstrapBlazor.Gantt/BootstrapBlazor.Gantt.bundle.scp.css';
@import '_content/BootstrapBlazor.IconPark/BootstrapBlazor.IconPark.bundle.scp.css';
@import '_content/BootstrapBlazor.ImageCropper/BootstrapBlazor.ImageCropper.bundle.scp.css';
@import '_content/BootstrapBlazor.Live2DDisplay/BootstrapBlazor.Live2DDisplay.bundle.scp.css';
@import '_content/BootstrapBlazor.MindMap/BootstrapBlazor.MindMap.bundle.scp.css';
@import '_content/BootstrapBlazor.MouseFollower/BootstrapBlazor.MouseFollower.bundle.scp.css';
@import '_content/BootstrapBlazor.SignaturePad/BootstrapBlazor.SignaturePad.bundle.scp.css';
@import '_content/BootstrapBlazor.Splitting/BootstrapBlazor.Splitting.bundle.scp.css';
@import '_content/BootstrapBlazor.SvgEditor/BootstrapBlazor.SvgEditor.bundle.scp.css';
@import '_content/BootstrapBlazor.Topology/BootstrapBlazor.Topology.bundle.scp.css';
@import '_content/BootstrapBlazor.WebAPI/BootstrapBlazor.WebAPI.bundle.scp.css';
@import '_content/BootstrapBlazor/BootstrapBlazor.bundle.scp.css';
```

你可以观察一下你的项目部署后的样式文件，是否缺少这些 `import`

**Comment by densen2014** [comment]
大概你用的不太清楚Densen informatica在 2024年9月29日，13:11，Argo Zhang ***@***.***> 写道：﻿
@WeiJunFenYou 我刚才想到一个问题点，应该是这个问题导致的
我猜你引用 SignaturePad 的工程未开启样式隔离，导致最终的样式隔离文件缺少 import 语句导致的
我们项目的样式文件如下：
@import '_content/BootstrapBlazor.AntDesignIcon/BootstrapBlazor.AntDesignIcon.bundle.scp.css';
@import '_content/BootstrapBlazor.Bluetooth/BootstrapBlazor.Bluetooth.bundle.scp.css';
@import '_content/BootstrapBlazor.CodeEditor/BootstrapBlazor.CodeEditor.bundle.scp.css';
@import '_content/BootstrapBlazor.ElementIcon/BootstrapBlazor.ElementIcon.bundle.scp.css';
@import '_content/BootstrapBlazor.Gantt/BootstrapBlazor.Gantt.bundle.scp.css';
@import '_content/BootstrapBlazor.IconPark/BootstrapBlazor.IconPark.bundle.scp.css';
@import '_content/BootstrapBlazor.ImageCropper/BootstrapBlazor.ImageCropper.bundle.scp.css';
@import '_content/BootstrapBlazor.Live2DDisplay/BootstrapBlazor.Live2DDisplay.bundle.scp.css';
@import '_content/BootstrapBlazor.MindMap/BootstrapBlazor.MindMap.bundle.scp.css';
@import '_content/BootstrapBlazor.MouseFollower/BootstrapBlazor.MouseFollower.bundle.scp.css';
@import '_content/BootstrapBlazor.SignaturePad/BootstrapBlazor.SignaturePad.bundle.scp.css';
@import '_content/BootstrapBlazor.Splitting/BootstrapBlazor.Splitting.bundle.scp.css';
@import '_content/BootstrapBlazor.SvgEditor/BootstrapBlazor.SvgEditor.bundle.scp.css';
@import '_content/BootstrapBlazor.Topology/BootstrapBlazor.Topology.bundle.scp.css';
@import '_content/BootstrapBlazor.WebAPI/BootstrapBlazor.WebAPI.bundle.scp.css';
@import '_content/BootstrapBlazor/BootstrapBlazor.bundle.scp.css';

你可以观察一下你的项目部署后的样式文件，是否缺少这些 import

**Reply by WeiJunFenYou** [reply]
看了下，的确没有这些import。微软的官方文档说，CSS 隔离开箱即用，不需要配置什么。大佬的项目中配置过与此有关的东西吗？

**Reply by ArgoZhang** [reply]
我不是告诉你了吗？认真读几遍我说的话，是我猜的。开启就好了

**Comment by WeiJunFenYou** [comment]
又找到了一个新的解决方法：
在页面中直接使用外链样式表
`<link href="/_content/BootstrapBlazor.SignaturePad/BootstrapBlazor.SignaturePad.bundle.scp.css" rel="stylesheet" />`
仍然治标不治本，但比拉取代码再打包的方式强很多。

**Reply by ArgoZhang** [reply]
下个里程碑，准备把这些样式统一处理一下，均采用自动加载的方式导入，与样式隔离技术脱钩，可以杜绝此类问题

**Reply by ArgoZhang** [reply]
@WeiJunFenYou @densen2014 发布 `8.0.3` 移除样式隔离，引用包即可，无需使用者处理样式或者脚本
